File: Script/Tools/SceneEditor/NodeCapabilities.lua
-- [ts]: NodeCapabilities.ts
local ____exports = {} -- 1
local ____NodeCatalog = require("Script.Tools.SceneEditor.NodeCatalog") -- 2
local getNodeKindDefinition = ____NodeCatalog.getNodeKindDefinition -- 2
local function nodeCapability(kind) -- 24
	return getNodeKindDefinition(kind) -- 25
end -- 24
function ____exports.isRootNodeKind(kind) -- 28
	return kind == "Root" -- 29
end -- 28
function ____exports.canDeleteNodeKind(kind) -- 32
	return nodeCapability(kind).canDelete == true -- 33
end -- 32
function ____exports.canNodeKindHaveChildren(kind) -- 36
	return nodeCapability(kind).canHaveChildren == true -- 37
end -- 36
function ____exports.canNodeKindBindTexture(kind) -- 40
	return nodeCapability(kind).canBindTexture == true -- 41
end -- 40
function ____exports.canNodeKindBindScript(kind) -- 44
	return nodeCapability(kind).canBindScript == true -- 45
end -- 44
function ____exports.canNodeKindEditText(kind) -- 48
	return nodeCapability(kind).canEditText == true -- 49
end -- 48
function ____exports.canNodeKindFollowTarget(kind) -- 52
	return nodeCapability(kind).canFollowTarget == true -- 53
end -- 52
function ____exports.canNodeKindBeFollowTarget(kind) -- 56
	return nodeCapability(kind).canBeFollowTarget == true -- 57
end -- 56
function ____exports.canSelectNodeKindInViewport(kind) -- 60
	return nodeCapability(kind).selectableInViewport == true -- 61
end -- 60
function ____exports.nodeKindViewportPickLayer(kind) -- 64
	return nodeCapability(kind).viewportPickLayer or "none" -- 65
end -- 64
function ____exports.nodeKindUsesTextureSize(kind) -- 68
	return nodeCapability(kind).usesTextureSize == true -- 69
end -- 68
function ____exports.nodeKindUsesGameFrameSize(kind) -- 72
	return nodeCapability(kind).usesGameFrameSize == true -- 73
end -- 72
function ____exports.defaultNodeText(kind) -- 76
	return nodeCapability(kind).defaultText or "" -- 77
end -- 76
function ____exports.defaultNodeVisualSize(kind) -- 80
	local capability = nodeCapability(kind) -- 81
	return {capability.defaultWidth or 72, capability.defaultHeight or 72} -- 82
end -- 80
function ____exports.shouldSpawnNodeAtOrigin(kind) -- 85
	return nodeCapability(kind).spawnAtOrigin == true -- 86
end -- 85
return ____exports -- 85

File: Script/Tools/SceneEditor/NodeCapabilities.ts
import { SceneNodeKind } from 'Script/Tools/SceneEditor/EditorTypes';
import { getNodeKindDefinition } from 'Script/Tools/SceneEditor/NodeCatalog';

export type ViewportPickLayer = 'object' | 'camera' | 'none';

interface NodeCapabilityDefinition {
	canDelete?: boolean;
	canHaveChildren?: boolean;
	canBindTexture?: boolean;
	canBindScript?: boolean;
	canEditText?: boolean;
	canFollowTarget?: boolean;
	canBeFollowTarget?: boolean;
	selectableInViewport?: boolean;
	viewportPickLayer?: ViewportPickLayer;
	usesTextureSize?: boolean;
	usesGameFrameSize?: boolean;
	spawnAtOrigin?: boolean;
	defaultText?: string;
	defaultWidth?: number;
	defaultHeight?: number;
}

function nodeCapability(kind: SceneNodeKind) {
	return getNodeKindDefinition(kind) as unknown as NodeCapabilityDefinition;
}

export function isRootNodeKind(kind: SceneNodeKind) {
	return kind === 'Root';
}

export function canDeleteNodeKind(kind: SceneNodeKind) {
	return nodeCapability(kind).canDelete === true;
}

export function canNodeKindHaveChildren(kind: SceneNodeKind) {
	return nodeCapability(kind).canHaveChildren === true;
}

export function canNodeKindBindTexture(kind: SceneNodeKind) {
	return nodeCapability(kind).canBindTexture === true;
}

export function canNodeKindBindScript(kind: SceneNodeKind) {
	return nodeCapability(kind).canBindScript === true;
}

export function canNodeKindEditText(kind: SceneNodeKind) {
	return nodeCapability(kind).canEditText === true;
}

export function canNodeKindFollowTarget(kind: SceneNodeKind) {
	return nodeCapability(kind).canFollowTarget === true;
}

export function canNodeKindBeFollowTarget(kind: SceneNodeKind) {
	return nodeCapability(kind).canBeFollowTarget === true;
}

export function canSelectNodeKindInViewport(kind: SceneNodeKind) {
	return nodeCapability(kind).selectableInViewport === true;
}

export function nodeKindViewportPickLayer(kind: SceneNodeKind) {
	return nodeCapability(kind).viewportPickLayer || 'none';
}

export function nodeKindUsesTextureSize(kind: SceneNodeKind) {
	return nodeCapability(kind).usesTextureSize === true;
}

export function nodeKindUsesGameFrameSize(kind: SceneNodeKind) {
	return nodeCapability(kind).usesGameFrameSize === true;
}

export function defaultNodeText(kind: SceneNodeKind) {
	return nodeCapability(kind).defaultText || '';
}

export function defaultNodeVisualSize(kind: SceneNodeKind): [number, number] {
	const capability = nodeCapability(kind);
	return [capability.defaultWidth || 72, capability.defaultHeight || 72];
}

export function shouldSpawnNodeAtOrigin(kind: SceneNodeKind) {
	return nodeCapability(kind).spawnAtOrigin === true;
}
